### PR TITLE
feat(builder): add delete handles to the edges and nodes

### PR DESCRIPTION
**Pull request by ntindle**
### Background

It's not clear how to delete edges and nodes at this time. This PR seeks to remedy that by adding clear deletion icons on hover to each

### Changes 🏗️

- Adds hover events to nodes and edges
- Adds delete button to edges
- increases stroke when an edge is hovered
- adds delete and (unhookedup) copy button to node

## Demo
image

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [x] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

___

### **PR Type**
Enhancement, Bug fix

___

### **Description**
- Added delete button to edges with hover effect and implemented edge deletion logic.
- Enhanced edge hover styling to increase stroke width.
- Added delete and copy buttons to nodes with hover effect.
- Implemented node deletion logic and added placeholder for node copy functionality.
- Styled delete and copy buttons for both edges and nodes, including hover and transition effects.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CustomEdge.tsx</strong><dd><code>Add delete button and hover effects to edges</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_builder/src/components/CustomEdge.tsx

<li>Added delete button to edges with hover effect<br> <li> Enhanced edge hover styling<br> <li> Implemented edge deletion logic<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7614/files#diff-42f43f21a4229cf0ed2f24d6e26260aba0b69a0aa505d18200096caa33bb407c">+52/-9</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CustomNode.tsx</strong><dd><code>Add delete and copy buttons to nodes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_builder/src/components/CustomNode.tsx

<li>Added delete and copy buttons to nodes with hover effect<br> <li> Implemented node deletion logic<br> <li> Added placeholder for node copy functionality<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7614/files#diff-ce37ba33e874cc42255fda0cd2a16758a94a6e68606d513b7018c1cc68193146">+97/-14</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>customedge.css</strong><dd><code>Style delete button for edges</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_builder/src/components/customedge.css

<li>Styled delete button for edges<br> <li> Added hover and transition effects for edge buttons<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7614/files#diff-04cf55605bb3db58d35e78cac2493e174915eae39d7862e5b2f45282b9a147fc">+38/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>customnode.css</strong><dd><code>Style delete and copy buttons for nodes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_builder/src/components/customnode.css

<li>Styled delete and copy buttons for nodes<br> <li> Added hover and transition effects for node buttons<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7614/files#diff-5bd354fa19e27aa2784e3ffea64648f1b39295a4507fb1d63f84e17cd8ec2b48">+55/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: rnd/autogpt_builder/src/components/CustomEdge.tsx [hunk 1]
```diff
@@ -25,12 +33,47 @@ const CustomEdgeFC: FC<EdgeProps<CustomEdgeData>> = ({ data, selected, source, s
...
+          }}
+          className="edge-label-renderer"
+        >
+          <button
+            onMouseEnter={() => setIsHovered(true)}
+            onMouseLeave={() => setIsHovered(false)}
+            className={`edge-label-button ${isHovered ? 'visible' : ''}`}
+            onClick={onEdgeClick}
+          >
+            <X size={14} />
```
Comment (andrewhooker2) [inline rnd/autogpt_builder/src/components/CustomEdge.tsx]: When working with lucide's icon's and displaying them on the FE  try leveraging classNames with tailwind. Using the size class from tailwind will set height and width.

```suggestion
            <X className="size-4" />
```

Comment (ntindle) [reply rnd/autogpt_builder/src/components/CustomEdge.tsx]: thats cool!

**Review by Bentlybro — APPROVED** [review]
tested and works well, very nice :)

**Review by Bentlybro — APPROVED** [review]
tested and works well, very nice :)

**Pull request merged**